docs/proposals/hpa/cronfederatedhpa.md

```diff
@@ -0,0 +1,347 @@
+
+## Design Details
+### Architecture
+<image src="./statics/cronfederatedhpa-architecture.png" width="60%">
+
+From the above architecture diagram, CronFederatedHPA should do following things:
+* `CronFederatedHPA` allow users to scale the workloads at specific time.
+* `CronFederatedHPA` allow users to set the `FederatedHPA` at specific time.
+
+### API Definition
+In order to satisfy the above requirements, we defined the API as following:
+```go
+type CronFederatedHPA struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the specification of the CronFederatedHPASpec.
+	// +required
+	Sepc CronFederatedHPASpec `json:"spec"`
+
+	// Status is the current status of the CronFederatedHPA.
+	// +optional
+	Status CronFederatedHPAStatus `json:"status"`
+}
+
+type CronFederatedHPASpec struct {
+	// ScaleTargetRef points to the target resource to scale.
+	// It could be FederatedHPA, Deployment, StatefulSet, or other resource with scale sub-resource.
+	// +required
+	ScaleTargetRef ScaleTargetRef `json:"scaleTargetRef"`
+
+	// Rules is the list of CronFederatedHPARule,
+	// it define when and how to scale the related resources.
+	Rules []CronFederatedHPARule `json:"rules"`
+}
+
+type CronFederatedHPARule struct {
+	// RuleName is the name of the CronFederatedHPARule.
+	// +required
+	RuleName string `json:"ruleName"`
+
+	// Disable is used to disable the CronFederatedHPARule.
+	// Default is false
+	// +optional
+	Disable bool `json:"disable"`
+
+	// Schedule is the schedule of the CronFederatedHPARule.
+	// The format should follow https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#schedule-syntax
+	// +required
+	Schedule string `json:"schedule"`
+
+	// TargetReplicas is the target replicas to be scaled for deployment, statefulset, etc.
+	// It should be set if you want to scale the related resources.
+	TargetReplicas *int32 `json:"targetReplicas,omitempty"`
+
+	// TargetMinReplicas is the target MinReplicas to be set for FederatedHPA.
+	// It should be set if you want to configure FederatedHPA.
+	// If you not set this field, the value will not be updated.
+	TargetMinReplicas *int32 `json:"targetMinReplicas,omitempty"`
+
+	// TargetMaxReplicas is the target MaxReplicas to be set for FederatedHPA.
+	// It should be set if you want to configure FederatedHPA.
+	// If you not set this field, the value will not be updated.
+	TargetMaxReplicas *int32 `json:"targetMaxReplicas,omitempty"`
+}
+
+type ScaleTargetRef struct {
+	// APIVersion of the referent resources.
+	ApiVersion string `json:"apiVersion"`
+
+	// Kind of the referent resources.
+	Kind string `json:"kind"`
+
+	// Name of the referent resources.
+	Name string `json:"name"`
+}
+
+type CronFederatedHPAStatus struct {
+	// Conditions is the current status of the CronFederatedHPA.
+	Conditions []Condition `json:"conditions"`
```

Comment:
I think we should use the standard `condition` that is `metav1.Condition`, like what we did at [ResourceBinding Conditions](https://github.com/karmada-io/karmada/blob/4e4bfffc61067ddbde6b287cdb338bea2f5647df/pkg/apis/work/v1alpha2/binding_types.go#LL265C15-L265C31).

Reply:
I carefully reviewed the code and found that the fields in `metav1.Condition` are insufficient to display the current status of each rule. Therefore, I suggest using a new struct to define the rule condition.

Reply:
For the status field, there are two options.
Option 1:
```yaml
apiVersion: autoscaling.karmada.io/v1alpha1
kind: CronFederatedHPA
metadata:
  name: cron-federated-hpa
spec:
  scaleTargetRef:
    apiVersion: autoscaling.karmada.io/v1alpha1
    kind: FederatedHPA
    name: shop-fhpa
  rules:
  - ruleName: "Scale-Up"
    schedule: "30 08 * * *"
    targetMinReplicas: 1000
  - ruleName: "Scale-Down"
    schedule: "0 11 * * *"
    targetMinReplicas: 1
status:
  ruleStates:
  - ruleName: "Scale-Up"
    nextExecutionTime: "2023-06-15T08:30:00Z"
    executionState: "Success"
    lastExecutionTime: "2023-06-14T08:30:00Z"
    message: "FederatedHPA(shop-fhpa) is scaled"
    targetMinReplicas: 1000
  - ruleName: "Scale-Down"
    nextExecutionTime: "2023-06-15T11:00:00Z"
    executionState: "Success"
    lastExecutionTime: "2023-06-14T11:00:00Z"
    message: "FederatedHPA(shop-fhpa) is scaled"
    targetMinReplicas: 1
```

Option two:
```yaml
apiVersion: autoscaling.karmada.io/v1alpha1
kind: CronFederatedHPA
metadata:
  name: cron-federated-hpa
spec:
  scaleTargetRef:
    apiVersion: autoscaling.karmada.io/v1alpha1
    kind: FederatedHPA
    name: shop-fhpa
  rules:
  - ruleName: "Scale-Up"
    schedule: "30 08 * * *"
    targetMinReplicas: 1000
  - ruleName: "Scale-Down"
    schedule: "0 11 * * *"
    targetMinReplicas: 1
status:
  ruleStates:
  - ruleName: "Scale-Up"
    nextExecutionTime: "2023-06-15T08:30:00Z"
    history:
    - executionState: "Success"
      executionTime: "2023-06-14T08:30:00Z"
      message: "FederatedHPA(shop-fhpa) is scaled"
      targetMinReplicas: 1000
    - executionState: "Failed"
      executionTime: "2023-06-13T08:30:00Z"
      message: "FederatedHPA(shop-fhpa) is not found"
      targetMinReplicas: 1000
  - ruleName: "Scale-Down"
    nextExecutionTime: "2023-06-15T11:00:00Z"
    history:
    - executionState: "Success"
      lastExecutionTime: "2023-06-14T11:00:00Z"
      message: "FederatedHPA(shop-fhpa) is scaled"
      targetMinReplicas: 1
    - executionState: "Failed"
      executionTime: "2023-06-13T11:00:00Z"
      message: "FederatedHPA(shop-fhpa) is not found"
      targetMinReplicas: 1
```

So which one is better? @Poor12 @XiShanYongYe-Chang @RainbowMango

Reply:
Option two